For control-plane-as-a-service, no probes will be created

unmarshall opened the pull request:
**How to categorize this PR?**
/kind enhancement
/area control-plane

**What this PR does / why we need it**:
Avoids creating probes for shoots which do not have any workers. 

**Which issue(s) this PR fixes**:
Fixes #80 and #76

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```other operator
Probes will not be created for shoots with no workers.
```

unmarshall commented:
/area control-plane

unmarshall commented:
/cc @rishabh-11

unmarshall commented:
/cc @acumino

shafeeqes commented on controllers/cluster_controller.go:
```diff
@@ -77,23 +77,34 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+	// if a shoot is created without any workers (this can only happen for control-plane-as-a-service use case), then if there is a probe registered then
+	// unregister the probe and return early. If there is no existing probe registered then return early.
+	if len(shoot.Spec.Provider.Workers) == 0 {
+		if r.ProberMgr.Unregister(req.Name) {
```
Why do we need to do this? Can't we just use a predicate to not requeue the request itself if the shoot doesn't have workers? Then the probe is never created.

unmarshall replied:
Calling `Unregister` is essentially a NoOp in this case since no probe with the shoot ns will be found. However, it allows me to support any future use cases (with no additional cost) where one could transition from workerless to shoot with workers and vice versa. As DWD i really do not have to then care if something transitions are supported or not.

shafeeqes replied:
Sure, but let's add a predicate as well and enqueue it only if there is a transition.

shafeeqes replied:
I am suggesting the predicate like this:
```
if newHasWorkers{
  return true
}else if oldHasWorkers{
  return true
}
return false
```

shafeeqes commented on controllers/cluster_controller.go:
```diff
@@ -64,7 +64,7 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	if notFound {
 		if r.ProberMgr.Unregister(req.Name) {
-			log.Info("Cluster not found, existing prober will be removed")
+			log.Info("Cluster not found, existing prober will be removed", "namespace", req.Namespace, "name", req.Name)
```
Doesn't the log contain these information already?

unmarshall replied:
Done

shafeeqes commented:
typo in PR name, `control-plane-as-a-service`

acumino commented on hack/addlicenseheaders.sh:
```diff
@@ -19,4 +19,13 @@ set -e
 
 # Uses the tool https://github.com/google/addlicense
-addlicense -v -c "SAP SE or an SAP affiliate company" -y 2022 -l apache -ignore "vendor/**" -ignore "**/*.md" -ignore "**/*.yaml" -ignore "**/Dockerfile" .
+YEAR=$1
+if [[ -z "$1" ]]; then
+  cat << EOF
+Unspecified 'YEAR' argument.
+Usage: addlicenceheaders.sh <YEAR>
+EOF
+  exit 1
+fi
```
Can we use date command here?

shafeeqes commented on cmd/probercmd.go:
```diff
@@ -18,7 +18,7 @@ import (
+	"github.com/gardener/dependency-watchdog/controllers/cluster"
 	"github.com/gardener/dependency-watchdog/internal/prober"
 	"github.com/gardener/dependency-watchdog/internal/util"
 	gardenextensions "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
```
call it `extensionsv1alpha1` or `gardenerextensionsv1alpha1`

unmarshall replied:
done

shafeeqes commented on controllers/cluster/cluster_controller.go:
```diff
@@ -153,7 +164,10 @@ func (r *ClusterReconciler) startProber(ctx context.Context, logger logr.Logger,
-		For(&gardenerv1alpha1.Cluster{}).
+		For(
+			&gardenerv1alpha1.Cluster{},
+			ctrlbuilder.WithPredicates(workerLessShoot(mgr.GetLogger())),
```
`mgr.GetLogger()` will not contain the controller name and kind. We should build the controller and use `c.GetLogger()` which is built with this info.
Ref: https://github.com/gardener/gardener/blob/b89462e43c1f56843d8d95bb1f459b1e0c0663b4/vendor/sigs.k8s.io/controller-runtime/pkg/builder/controller.go#L327-L336

unmarshall replied:
done

shafeeqes commented:
/test pull-dependency-watchdog-unit

shafeeqes commented on controllers/endpoint/endpoints_controller.go:
```diff
@@ -57,20 +63,32 @@ func (r *EndpointReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
-		WithOptions(controller.Options{MaxConcurrentReconciles: r.MaxConcurrentReconciles}).
-		Complete(r)
+			MatchingEndpoints(r.WeederConfig.ServicesAndDependantSelectors),
+			ReadyEndpoints(mgr.GetLogger()),
```
```suggestion
			ReadyEndpoints(c.GetLogger()),
```

unmarshall replied:
This has been changed. PTAL.

unmarshall commented on controllers/cluster/cluster_controller.go:
```diff
@@ -151,9 +167,16 @@ func (r *ClusterReconciler) startProber(ctx context.Context, logger logr.Logger,
+	if err != nil {
+		return err
+	}
+	return c.Watch(&source.Kind{Type: &gardenerv1alpha1.Cluster{}}, &handler.EnqueueRequestForObject{}, workerLessShoot(c.GetLogger()))
```
done. PTAL

unmarshall commented:
@acumino, @shafeeqes I have now addressed all your review comments. PTAL.
@rishabh-11, @himanshu-kun please have a look at from DWD perspective and see if you want any change.

shafeeqes commented on controllers/endpoint/endpoints_controller.go:
```diff
@@ -26,12 +26,18 @@ import (
 
-// EndpointReconciler reconciles an Endpoints object
-type EndpointReconciler struct {
+const (
+	controllerName = "endpoint"
+)
```
Nit:
```suggestion
const controllerName = "endpoint"
```

unmarshall replied:
its done in both the controllers now. PTAL

shafeeqes reviewed (commented):
Thanks for addressing my comments.
One minor comment.
Otherwise lgtm

shafeeqes reviewed (commented):
/lgtm

himanshu-kun reviewed (approved):
Thanks for the effort , the structure is much cleaner now!
Just a nit 
approving anyways

/lgtm
